GEOSEARCH bybox bug fixes and new fuzzy tester

yangbodong22011 opened the pull request:
[Edit]

- Fix GEOSEARCH bybox width and height mismatch
- Fix errors of GEOSEARCH bybox search due to projection of the box to a
  trapezoid (when the meter box is converted to long / lat it's no longer a box).
- Add GEOSEARCH bybox testing to the existing "GEOADD + GEORANGE randomized test"
- Add new fuzzy test to stress test the bybox corners and edges
- Add some tests for edge cases of the bybox algorithm


---

This PR add fuzzy test for `GEOSEARCH` command and fix the wrong order of widht and height parameters. As shown in the [document](https://redis.io/commands/geosearch), width is before height.

oranagra commented on src/geo.c:
```diff
@@ -176,8 +176,8 @@ int extractDistanceOrReply(client *c, robj **argv,
-    if ((getDoubleFromObjectOrReply(c, argv[0], &h, "need numeric height") != C_OK) ||
-        (getDoubleFromObjectOrReply(c, argv[1], &w, "need numeric width") != C_OK)) {
+    if ((getDoubleFromObjectOrReply(c, argv[0], &h, "need numeric width") != C_OK) ||
+        (getDoubleFromObjectOrReply(c, argv[1], &w, "need numeric height") != C_OK)) {
```
looks like you only fixed the error message, but not the actual parameters
```suggestion
    if ((getDoubleFromObjectOrReply(c, argv[0], &w, "need numeric width") != C_OK) ||
        (getDoubleFromObjectOrReply(c, argv[1], &h, "need numeric height") != C_OK)) {
```

oranagra commented on src/geo.c:
```diff
@@ -176,8 +176,8 @@ int extractDistanceOrReply(client *c, robj **argv,
 int extractBoxOrReply(client *c, robj **argv, double *conversion,
                          double *height, double *width) {
```
you had two more bugs, canceling one another.
```suggestion
int extractBoxOrReply(client *c, robj **argv, double *conversion,
                         double *width, double *height) {
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -1,6 +1,7 @@
 # verify the Redis implementation with a fuzzy test.
-proc geo_degrad deg {expr {$deg*atan(1)*8/360}}
+proc geo_degrad deg {expr {$deg*(atan(1)*8/360)}}
+proc geo_raddeg deg {expr {$deg/(atan(1)*8/360)}}
```
```suggestion
proc geo_raddeg rad {expr {$rad/(atan(1)*8/360)}}
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -42,6 +43,31 @@ proc compare_lists {List1 List2} {
+    return false
+}
+
+# If a point in box
```
please add a comment explaining fixed_degree

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -42,6 +43,31 @@ proc compare_lists {List1 List2} {
+    set width_m [expr {$width_km*1000/2}]
+    set height_m [expr {$height_km*1000/2}]
+    set long_delta [geo_raddeg [expr {$width_m/6372797.560856/cos([geo_degrad $search_lat])}]]
+    set lat_delta [geo_raddeg [expr {$height_m/6372797.560856}]]
```
it would be nice to explain this (spherical projection? earth radius?)

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -374,53 +402,84 @@ start_server {tags {"geo"}} {
+                    set width_km [expr {[randomInt 5000]+10}]
+                    set height_km [expr {[randomInt 5000]+10}]
+                }
                 # From time to time use very big radiuses
```
IMHO that comment should be above.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -374,53 +402,84 @@ start_server {tags {"geo"}} {
                 # Normally use a few - ~200km radiuses to stress
                 # test the code the most in edge cases.
-                set radius_km [expr {[randomInt 200]+10}]
+                if {$type == "byradius"} {
+                    set radius_km [expr {[randomInt 200]+10}]
+                } elseif {$type == "bybox"} {
+                    set width_km [expr {[randomInt 200]+10}]
+                    set height_km [expr {[randomInt 200]+10}]
+                }
+            }
+            if {$rv ne {}} {
+                set radius_km [lindex $rv 1]
+                set width_km [lindex $rv 1]
+                set height_km [lindex $rv 1]
```
i have a feeling that it would be cleaner to just have these converted to meters, and don't mention `km` anywhere in the code (except for the comment above).

oranagra commented:
@yangbodong22011 thank you for this PR.

this is not exactly the test i wanted, but maybe its is enough.
correct me if i'm wrong, this test just checks that the same calculations performed by redis and by the tcl code reach the same conclusion (and it does so with random coordinates.

what i was looking is more of a specific test that tests edge cases in the box code.
IIRC our implementation does something like this (converting a box into a circle):
![image](https://user-images.githubusercontent.com/7045099/105414181-d3bce880-5c3f-11eb-820c-fed042bfc5f4.png)
i would like to test the 4 (or maybe just 2) green dots, and maybe the orange ones too, to see that they are handled correctly.

and i would like to do that on various different box shapes (horizontally wide, and horizontally narrow).
and i would like to places these boxes in various places on the globe (i.e. since our search grid is in a fixed positions, some bugs can reproduce on a certain long/lat position and not on another)

i didn't dig too much into the current test, if you can convince me that it covers the above concern, that would be great.
thank you.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -360,12 +386,14 @@ start_server {tags {"geo"}} {
-    test {GEOADD + GEORANGE randomized test} {
+    foreach {type} {byradius bybox} {
+    test "GEOSEARCH fuzzy test - $type" {
         set attempt 30
```
```suggestion
        if {$::accurate}{ set attempt 300 } else { set attempt 30 }
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -374,53 +402,84 @@ start_server {tags {"geo"}} {
+                        }
+                    } elseif {$type == "bybox"} {
+                        # we add fixed 0.01 degree because higher precision calculation
+                        if {[pointInBox $width_km $height_km $lon $lat $search_lon $search_lat 0.01]} {
```
IIUC, the original code is adding 1 meter (0.999 KM), and you're adding 0.01 degree which could be as much as 1km, right?

on the other hand, converting a box in km (square) to a lat / long box at a certain latitude is not an accurate business (no longer a box in km), which makes me wonder if what we're doing here is right?

yangbodong22011 commented:
Updated:

- Fixed the error in latitude and introduced a trapezoid. To determine whether a point is in the trapezoid, use the ray-crossing Algorithm, reference: `geohashGetDistanceIfInTrapezoid()`

- Add tcl test, in tcl, to judge whether a point is in the trapezoid, use another area Algorithm, reference: `pointInBox()`, it's different from c code.

- Added a test that the point is very close to the box

NOTICE:
- shape->bounds now are the 4 coordinates of the trapezoid.
- use sqrt((width/2)^2 + (height/2)^2) for radius_meters, because in some case, max(height/2, widht/2) is not enough.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -54,18 +54,41 @@ proc pointInCircular {radius_km lon lat search_lon search_lat} {
+    }
+
+    if {$cross % 2 == 1} {
+        return true;
```
`;` is excessive

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -54,18 +54,41 @@ proc pointInCircular {radius_km lon lat search_lon search_lat} {
-        $lat < [expr {$search_lat-$lat_delta-$fixed_degree}] ||
-        $lat > [expr {$search_lat+$lat_delta+$fixed_degree}]} {
-        return false
+    set long_delta_top [geo_raddeg [expr {$width_m/6372797.560856/cos([geo_degrad [expr {$search_lat+$lat_delta}]])}]]
```
maybe cache top_lat and bottom_lat before all of that, the code be a bit more clear.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -54,18 +54,41 @@ proc pointInCircular {radius_km lon lat search_lon search_lat} {
 
 # If a point in box
-proc pointInBox {width_km height_km lon lat search_lon search_lat fixed_degree} {
+proc pointInBox {width_km height_km lon lat search_lon search_lat} {
```
let's add some comment in the lines of:
```
convert width and height to half width and half height in meters.
convert to 4 long/lat coordinates using the earth radius and the center point.
check if point is inside the rectangle
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -476,10 +499,7 @@ start_server {tags {"geo"}} {
                         }
                     } elseif {$type == "bybox"} {
                         # we add fixed 0.01 degree because higher precision calculation
```
the comment is outdated

yangbodong22011 replied:
The current code does not have this comment. I don’t know why you can still comment here.

oranagra replied:
probably because i reviewed the 3 commits one by one, so i added this comment while reviewing the first commit.

oranagra commented on src/geohash_helper.c:
```diff
@@ -83,35 +83,38 @@ uint8_t geohashEstimateStepsByRadius(double range_meters, double lat) {
- * Since this function is currently only used as an optimization, the
- * optimization is not used for very big radiuses, however the function
- * should be fixed. */
+ * (bounds[0],bounds[1])   (bounds[2],bounds[3])
+ *             \-----------------/                   --------
+ *              \               /                  /          \
+ *               \  (long,lat) /                  / (long,lat) \
+ *                \           /                  /              \
+ *                  ---------                   /----------------\
+ *(bounds[6],bounds[7])   (bounds[4],bounds[5])
+ *               Northern Hemisphere              Southern Hemisphere
```
a box that's very high and narrow and is placed around the equator will actually look like this:
```
*             \-----------------/ 
*              \               / 
*               \  (long,lat) / 
*                /            \
*              /                \
*             /------------------\
```
maybe we should support that?
besides, the top and bottom edges are actually bent (not straight)

yangbodong22011 replied:
yes !  you are right, so we may need save 6 coordinates.  (long_delta_top, long_delta_middle, long_delta_bottom ) ,although their sides are not straight lines, we use these 6 points as bounds. Whether in the northern and southern hemispheres or across the equator.

oranagra replied:
it's not the sides that are not straight lines, it's the top and bottom.
```
 *              ___---`````---___
 *              \               /    
 *               \  (long,lat) / 
 *                \           /      
 *                  ---```---      
```

oranagra commented on src/geohash_helper.c:
```diff
@@ -128,21 +131,23 @@ GeoHashRadius geohashCalculateAreasByShapeWGS84(GeoShape *shape) {
-    max_lat = shape->bounds[3];
+    /* The trapezoid directions of the northern and southern hemispheres
+     * are opposite, so we choice different points as max/min long/lat*/
+    int southern_hemisphere = shape->xy[1] < 0 ? 1 : 0;
```
as noted above, it can be in both the north and south, since it's a box, if the center is slightly above the equator, then indeed the north part is wider, so i suppose this code can work as is (even if we fix it to no longer be a trapezoid)

oranagra commented on src/geohash.h:
```diff
@@ -96,8 +96,8 @@ typedef struct {
     double conversion; /* km: 1000 */
-    double bounds[4]; /* bounds[0]: min_lon, bounds[1]: min_lat
-                       * bounds[2]: max_lon, bounds[3]: max_lat */
+    double bounds[8]; /* bounds[0] - bounds[7] are the 4 coordinates of the trapezoid
```
i suppose we should rename this ("bounds" is no longer a good term)?

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -535,4 +539,42 @@ start_server {tags {"geo"}} {
     } {OK}
     }
+
+    test {GEOSEARCH fuzzy test} {
```
that's a good test, but i think we need to extend it.
currently you only have `lat_delta`, so you only test the top edge of the trapezoid.
i think you should at the very least test also:
1. a horizontal edge (i.e. have a `long_delta`
2. try to apply both at the same time, so that you test the corners.

ideally, we should test:
1. a point inside and a point outside at the center of each of the edges. (total of 8 tests)
2. 4 points around each of the corners (total of 16 tests).

but it may be sufficient to just test one corner, and two edges (horizontal one and vertical one).

yangbodong22011 replied:
update: 
The latest code generates 8 points, which are located at each vertex and the center of each side.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -535,4 +539,42 @@ start_server {tags {"geo"}} {
+            r del mypoints
+
+            geo_random_point search_lon search_lat
+            set width_m [expr {[randomInt 1000]+10}]
+            set height_m [expr {[randomInt 1000]+10}]
```
maybe we should have some mechanism to randomly also choose very big boxes, or very long and narrow ones.

yangbodong22011 replied:
The current code also satisfies the narrow and long condition. Although the current box length and width are limited to 1000 meters, the mobile box only moves 1 meter, so it can maintain a 0.1% error rate.

Although 1,000 meters is not large enough, it is long and narrow for 1 meter.

oranagra reviewed (commented):
@yangbodong22011 thank you!
i assume the improved tests failed quite consistently on the old implementation, right?

considering the equator problem i mentioned and maybe the bent edges, they should still fail from time to time, and we can improve both tests and the code.

yangbodong22011 commented:
> Updated:
> 
> * Fixed the error in latitude and introduced a trapezoid. To determine whether a point is in the trapezoid, use the ray-crossing Algorithm, reference: `geohashGetDistanceIfInTrapezoid()`
> * Add tcl test, in tcl, to judge whether a point is in the trapezoid, use another area Algorithm, reference: `pointInBox()`, it's different from c code.
> * Added a test that the point is very close to the box
> 
> NOTICE:
> 
> * shape->bounds now are the 4 coordinates of the trapezoid.
> * use sqrt((width/2)^2 + (height/2)^2) for radius_meters, because in some case, max(height/2, widht/2) is not enough.

- tcl test also use ray-crossing Algorithm, because the area algorithm is not suitable for concave polygons.
- update fuzzy test, now 8 points are randomly generated.
- When the trapezoid crosses the equator, the equator is selected as the narrowest.
- When the box spans -180° or 180°, the searched point + or-360° is required, because `180° + 1° = -179°`

@oranagra please have a look when you have time.

oranagra commented on src/geohash_helper.c:
```diff
@@ -128,21 +146,23 @@ GeoHashRadius geohashCalculateAreasByShapeWGS84(GeoShape *shape) {
+     * by a circle. */
     double radius_meters = shape->type == CIRCULAR_TYPE ? shape->t.radius :
-                            shape->t.r.width > shape->t.r.height ? shape->t.r.width/2 : shape->t.r.height/2;
+            sqrt((shape->t.r.width/2)*(shape->t.r.width/2) + (shape->t.r.height/2)*(shape->t.r.height/2));
```
by now, can't we just calculate the distance from the center to b[0],b[1]?
or actually max between the distance of the center to b[0],b[1] and the distance of the center to b[6],b[7]?
i.e measure distance to the corners instead of a sqrt?

yangbodong22011 replied:
> by now, can't we just calculate the distance from the center to b[0],b[1]?

yes, we can, the distance from the center to b[0],b[1] and sqrt are equal.

oranagra replied:
Ohh, actually that involves some cosine right? not sure which one would be more expensive to compute or more accurate. 

main difference is that one is based on meters, and the other on the translated lat/lon of the corners.

oranagra commented on src/geohash_helper.c:
```diff
@@ -245,14 +265,94 @@ int geohashGetDistanceIfInRadiusWGS84(double x1, double y1, double x2,
+ * ray-crossing Algorithm refer: http://erich.realtimerendering.com/ptinpoly/
  */
-int geohashGetDistanceIfInRectangle(double *bounds, double x1, double y1,
+int geohashGetDistanceIfInTrapezoid(double *bounds, double x1, double y1,
```
this is no longer a trapezoid, right? it's a 6 vertex polygon. so let's rename the function.
come to think of it the `bounds` argument is no longer representing bounds, so it should be renamed too, right?
in that case, both "trapezoid" and "bounds" should be renamed to "polygon"?

i suppose nearly all mentions of "trapezoid" should be replaces with "polygon" except for the one next to the ascii art drawing.
and i suppose that the bounds member that holds 16 doubles (8 points) should be renamed to "polygon" too.
if we do need min/max lon/lat as real bounds (for radius search too), they should be stored in a different members, and not mixed with the polygon coordinates.

yangbodong22011 replied:
change to `geohashGetDistanceIfInPolygon`

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -42,6 +43,81 @@ proc compare_lists {List1 List2} {
+}
+
+# Judge whether a point is in the trapezoid.
+proc pointInBox {width_km height_km lon lat search_lon search_lat error} {
```
there's a contradiction between the comment and the function name.
maybe the comment should be removed from above the function (at this point it's still a box, not a trapezoid), 

and other comments placed inside it near where the `bounds` are calculated, that explain that now it's no longer a box, but a trapezoid or a polygon.

instead we need a top comment that says which of the lon/lat arguments is the box center, and which one is the point being searched. (similarly to the suggestion i made for pointInCircular)

and again, i think the terms "bounds" and "trapezoid" are not good (should be "cooords" and "polygon"

yangbodong22011 replied:
updated.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -42,6 +43,81 @@ proc compare_lists {List1 List2} {
+    set bounds(8) [expr {$search_lon-$long_delta_bottom}]
+    set bounds(9) [expr {$search_lat-$lat_delta}]
+    set bounds(10) [expr {$search_lon-$long_delta_middle}]
+    set bounds(11) $search_lat
```
this test is a bit weak since it uses the same approach as the code it tests.
luckily we have the other new test (one that shifts the box by 1m) that covers for this.
but maybe we can improve this function too.

some wild idea is to try to convert the point that we're matching to the coordinate system of the box (km) rather than converting the box to long/lat.
consider:
* the search box center being: lon1,lat1 (`search_lon` and `search_lat` right?)
* the the point we're to match to the box: lon2,lat2 (`lon` and `lat`, right?)

we can create two new points, and measure the horizontal and vertical distance from the box center and match them to width and height.
```
# on one longitude (the one of the point being searched),
# measure the distance between two lats (the lat of the point being search and the one of the center of the box)
distance between lon2/lat1 and lon2,lat2 < height/2 ?

# on one latitude (the one of the point being searched),
# measure the distance between two lons (the lon of the point being search and the one of the center of the box)
distance between lon1,lat2 and lon2,lat2 < width/2 ?
```
is this silly? (it's very late at night here), or could that work?

oranagra replied:
@yangbodong22011 what do you think of this?
if that algorithm indeed works, maybe instead of using it in the tests we should use it in redis itself?
i initially suggested it just so that we don't use the same algo in the tests as the one in redis (weak test), but if we prove that it's a valid algo, it might be more accurate than the polygon one.

yangbodong22011 replied:
@oranagra Thank you very much for your algorithm, i finished it at: https://github.com/redis/redis/pull/8445

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -42,6 +43,81 @@ proc compare_lists {List1 List2} {
 }
 
+# If a point in circular
+proc pointInCircular {radius_km lon lat search_lon search_lat} {
```
```suggestion
# return true If a point in circle.
# search_lon and search_lat define the center of the circle,
# and lon, lat define the point being searched.
proc pointInCircle {radius_km lon lat search_lon search_lat} {
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -42,6 +43,81 @@ proc compare_lists {List1 List2} {
+    }
+    if {$bounds(0) < -180 || $bounds(8) < -180} {
+        if {$lon > 0} {set lon [expr {$lon-360}]}
+    }
```
if we do keep the current code, please add a short comment above each block describing what it does

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +579,74 @@ start_server {tags {"geo"}} {
     } {OK}
+    }
+
+    test {GEOSEARCH fuzzy test} {
```
```suggestion
    test {GEOSEARCH box edges fuzzy test} {
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +579,74 @@ start_server {tags {"geo"}} {
+            set long_delta_bottom [geo_raddeg [expr {$width_m/2/6372797.560856/cos([geo_degrad [expr {$search_lat-$lat_delta}]])}]]
+
+            # Total of 8 points are generated, which are located at each vertex and the center of each side
+            set bounds(0) [expr {$search_lon-$long_delta_top}]
```
`bounds` is a bad term. maybe `points`?
p.s. it would be nicer to use a two dimensional array (nested lists) rather than a flat list of lon/lat pairs.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +579,74 @@ start_server {tags {"geo"}} {
+
+            set res2 [lsort $tcl_result]
+
+            # move the box by two meter to put the coordinate slightly inside the box.
```
```suggestion
            # make the box larger by two meter in each direction to put the coordinate slightly inside the box.
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +579,74 @@ start_server {tags {"geo"}} {
+                fail "place should be found, debuginfo: $debuginfo, height_new: $height_new width_new: $width_new"
+            }
+
+            # move the box by two meter to put the coordinate slightly outside the box.
```
```suggestion
            # make the box smaller by two meter in each direction to put the coordinate slightly outside the box.
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +579,74 @@ start_server {tags {"geo"}} {
+            }
+
+            # move the box by two meter to put the coordinate slightly outside the box.
+            set height_new [expr {$height_m-4}]
+            set width_new [expr {$width_m-4}]
```
for the 4 corner coordinates, i'd like to have two tests. one which shrinks only with width and one which shrinks only the height. in both cases the point is outside the box (now you test that when you shrink both width and height the point is outside the box).

maybe we can convert this block to two optional tests (i.e. after the one above that makes the box larger in all directions). so this block below will be something like:

```
if the lat of the point being search matches the lat of the top or bottom then
    run a test that shrinks the height
if the lon of the point being search matches the lon of the left or right then
    run a test that shrinks the width
```
this way, only one of these two tests will be processed on the 4 points that are in the center of each edge, but both tests will be processed on each of the 4 corners.

yangbodong22011 replied:
In fact, we currently reduce both height and width, which can cover the situation you mentioned (only reduce height or width), but I added the test case you mentioned.

oranagra replied:
the reason i didn't like the previous test was that if we have a bug in the corners handing, and that bug only affects the longitude, the test wouldn't have found the bug.
i.e. because reducing the height would have moved the point outside the search area, and we wouldn't have known that reducing the width would have still found it.

oranagra commented on src/geohash_helper.c:
```diff
@@ -83,41 +83,59 @@ uint8_t geohashEstimateStepsByRadius(double range_meters, double lat) {
+ *            (bounds[8],bounds[9])   (bounds[6],bounds[7])
+ *                       Northern Hemisphere                        Southern Hemisphere         Around the equator
+ */
 int geohashBoundingBox(GeoShape *shape, double *bounds) {
```
i think this function lost it's purpose, and that we're doing some unnecessary overheads in case of a search with radius.
the function is called "bounding box" and it writes to "bounds" variable.
but the bounds variable does't really holds bounds. it can hold a polygon.
i think i'd like to add min_lat, max_lat, min_lon, max_lon members to the shape struct, so this function will populate them and fulfill it's name.
and i think we should have a polygon member in the shape struct (16 doubles) that's only used in case of a box search (don't wanna run that overhead when working with a radius)

yangbodong22011 replied:
updated.

yangbodong22011 commented:
@oranagra Thank you very much for your detailed reply. I have updated the code, please have a look when you have time.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +589,92 @@ start_server {tags {"geo"}} {
+            set long_delta_middle [geo_raddeg [expr {$width_m/2/6372797.560856/cos([geo_degrad $search_lat])}]]
+            set long_delta_bottom [geo_raddeg [expr {$width_m/2/6372797.560856/cos([geo_degrad [expr {$search_lat-$lat_delta}]])}]]
+
+            # Total of 8 points are generated, which are located at each vertex and the center of each side
+            set points(0) [expr {$search_lon-$long_delta_top}]
+            set points(1) [expr {$search_lat+$lat_delta}]
+            set points(2) [expr {$search_lon+$long_delta_top}]
+            set points(3) [expr {$search_lat+$lat_delta}]
+            set points(4) [expr {$search_lon+$long_delta_bottom}]
+            set points(5) [expr {$search_lat-$lat_delta}]
+            set points(6) [expr {$search_lon-$long_delta_bottom}]
+            set points(7) [expr {$search_lat-$lat_delta}]
+            set points(8) $search_lon
+            set points(9) [expr {$search_lat+$lat_delta}]
+            set points(10) $search_lon
+            set points(11) [expr {$search_lat-$lat_delta}]
+            set points(12) [expr {$search_lon+$long_delta_middle}]
+            set points(13) $search_lat
+            set points(14) [expr {$search_lon-$long_delta_middle}]
+            set points(15) $search_lat
+
+            lappend debuginfo "Search area: geosearch mypoints fromlonlat $search_lon $search_lat bybox $width_m $height_m m"
+            set tcl_result {}
+            for {set i 0} {$i < 16} {incr i 2} {
+                set x $points($i)
+                set y $points([expr {$i+1}])
+                r geoadd mypoints $x $y place:$i
+                lappend tcl_result "place:$i"
+                lappend debuginfo "geoadd mypoints $x $y place:$i"
+            }
```
i think that instead of this long list of numbers (some represent long and some represent lat), and then a loop that work on pairs pf consecutive indexes), it would be nicer to create a proc that takes name lon lat adds a point.
this way, we can name them "north", "south", "north-west", "north-east", etc, and the code that skips points below can be clearer (not referring to "place:12" and "place:14"

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +589,92 @@ start_server {tags {"geo"}} {
+                fail "place should be found, debuginfo: $debuginfo, height_new: $height_new width_new: $width_new"
+            }
+
+            # Only width reduced, the four corners and the midpoint on the left and right will not be searched
```
did you mean?
```suggestion
            # Only width reduced, the four corners and the midpoint on the left and right will not be found
```

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +589,92 @@ start_server {tags {"geo"}} {
+                fail "place should not be found, debuginfo: $debuginfo, height_new: $height_new width_new: $width_new"
+            }
+
+            # Only height reduced, the four corners and the midpoint of the top and bottom will not be searched
```
did you mean?
```suggestion
            # Only height reduced, the four corners and the midpoint of the top and bottom will not be found
```

oranagra commented on src/geohash_helper.c:
```diff
@@ -82,69 +82,101 @@ uint8_t geohashEstimateStepsByRadius(double range_meters, double lat) {
+    if (shape->type == CIRCULAR_TYPE) {
+        radius_meters = shape->t.radius * shape->conversion;
+    } else if (shape->type == RECTANGLE_TYPE) {
+        radius_meters = geohashGetDistance(longitude, latitude, shape->polygon[0], shape->polygon[1]);
```
ohh, i see `geohashGetDistance` also has sqrt, so this approach may be slower than the previous formula (this one has a few `sin`, `cos` and `asin` ).
we can keep it if we believe this approach may be more accurate.

besides, i'm worried that maybe the distance from the center to the top,left vertex, may be in some cases smaller than the distance to the bottom right (depending on which one is closer the the pole).

what do you think? revert?

yangbodong22011 replied:
oh, I did not notice the sqrt of geohashGetDistance before, i will revert.

oranagra commented on tests/unit/geo.tcl:
```diff
@@ -456,4 +589,92 @@ start_server {tags {"geo"}} {
+            # Only width reduced, the four corners and the midpoint on the left and right will not be searched
+            set width_new [expr {$width_m-4}]
+            set res [lsort [r geosearch mypoints fromlonlat $search_lon $search_lat bybox $width_new $height_m m]]
+            if {$res != {place:10 place:8}} {
```
doesn't this make the test possibly flaky? i.e. the two points that are found are right on the border, due to inaccuracy issues they may be found time times and not found other times?
so we know the corners and top and right midpoints will not be found, but we're not sure about the left and right.

maybe one way to keep the current `if {$res != {place:10 place:8}} {` and make it stable is change the two tests to increase one dimension and decrease the other.
in this case we'll have 3 tests:
1. increase both width and height (all points are found)
2. increase width and decrease height (corners and top, bottom edges are found. only left and right)
3. decrease the width and increase the height (corners and left, right edges are found. only top and bottom)
4. if we want we can decrease both and make sure nothing is found but i don't think this test has value.

oranagra replied:
sorry.. ignore that suggestion of increasing width and decreasing height (the corners will not be found).
let's just consider the accuracy of whether place:10 and place:8 must be found (think sometimes they won't be found, and that's not an error)

oranagra commented:
closed in favor of #8445
